### PR TITLE
review

### PR DESCRIPTION
**Pull request by ZhiHanZ**
code review

### CODE REVIEW

File: service/service_helper.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
+#ifndef SERVICE_HELPER_H_
+#define SERVICE_HELPER_H_
+
+#include "service.pb.h"
+#include "unique_id.h"
+#include <atomic>
+#include <chrono>
```
Comment (ZhiHanZ) [inline service/service_helper.h]: <chrono> is an unapproved C++11 header. 
 <mutex> is an unapproved C++11 header.?

File: command/command.cc [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  }
+  return status;
+}
+// Post a chirp
+Chirp CommandClient::ChirpPost(const string &username, const string &chirp,
+                               const string &reply) {
+  ClientContext context;
+  auto request = ChirpRequestMaker(username, chirp, reply);
+  auto chirply = new ChirpReply;
+  if (reply != "-1") {
```
Comment (barath) [inline command/command.cc]: Move -1 to a constant.

File: command/command_helper.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+DEFINE_string(chirp, "", "a new chirp with the given text");
+DEFINE_string(reply, "", "please denote a chirp id you want to reply");
+DEFINE_string(follow, "", "please enter a username you want to follow");
+DEFINE_string(read, "", "read the chirp thread at a given id");
+DEFINE_bool(monitor, false, "please enter the user name you want to monitor");
+DEFINE_bool(lt, true, "little relation");
+const static string VOIDREGISTERERROR("username should contain at least one chararacter");
+const static string SUCCESS("success");
+const static string FAIL("some thing is wrong");
+namespace helper {
```
Comment (barath) [inline command/command_helper.h]: Add blank line.

File: command/command_helper.h [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  static auto ChirpRequestMaker(const string& registeruser,
+                                const string& chirp, 
+                                const string& reply){
+    ChirpRequest* request = new ChirpRequest;
+    request->set_username(registeruser);
+    request->set_text(chirp);
+    request->set_parent_id(reply);
+    return request;
+  }
+  //Make follow request
```
Comment (barath) [inline command/command_helper.h]: Add blank line (here and generally between declarations).

File: kvstore/cchash.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    GetBucket(key).AddOrUpdateValue(key, val);
+  }
+  // delete key.
+  void DeleteKey(const Key &key) { GetBucket(key).DeleteKey(key); }
+
+private:
+  // do not allow copy operation
+  ConcurrentHashTable(const ConcurrentHashTable &other) = delete;
+  auto &operator=(const ConcurrentHashTable &other) = delete;
+  class BucketType {
```
Comment (barath) [inline kvstore/cchash.h]: In addition to separating, add a class comment.

File: kvstore/cchash.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      if (FindEntry(key) != data_.end()) {
+        data_.erase(FindEntry(key));
+      }
+    }
+
+  private:
+    typedef std::pair<Key, Value> bucket_value;
+    typedef std::list<bucket_value> bucket_data;
+    bucket_data data_;
+    typedef typename bucket_data::iterator bucket_iter;
```
[the remark this answers is not in the document]
Comment (barath) [reply kvstore/cchash.h]: It's ok, but it seems a bit unnecessary.

File: kvstore/cchash.h [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    mutable std::shared_mutex mutex; // c++ 17 feature
+    bucket_iter FindEntry(const Key &key) {
+      return std::find_if(
+          data_.begin(), data_.end(),
+          [&key](bucket_value const &item) { return item.first == key; });
+    }
+  };
+  std::vector<std::unique_ptr<BucketType>> table_;
```
Comment (barath) [inline kvstore/cchash.h]: Add member comments.

File: kvstore/cchash.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  Hash hash_;
+  const static unsigned default_buckets_ = 19;
```
Comment (barath) [inline kvstore/cchash.h]: Constant should be named following the convention.

File: kvstore/kvhelper.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  return table.GetValue(key);
+}
+// Delete value through key
+// 0 represents delete successful
+//-1 represents delete a key do not exists
+int64_t Delete(ConcurrentHashTable<string, string> &table, const string &key) {
+  if (table.Has(key)) {
+    table.DeleteKey(key);
+    return 0;
+  } else {
```
Comment (barath) [inline kvstore/kvhelper.h]: Else not needed.

File: kvstore/kvstore.cc [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        return Status(StatusCode::ALREADY_EXISTS, "key does not exist.");
+      }
+    }
+  }
+  // delete a key . if key do not exists, return a error
+  Status deletekey(ServerContext *context, const DeleteRequest *request,
+                   DeleteReply *reply) override {
+    if (Delete(database, request->key()) == 0) {
+      return Status::OK;
+    } else {
```
Comment (barath) [inline kvstore/kvstore.cc]: Else not needed.

File: service/service.cc [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+  queue.push(*root);
+  while (!queue.empty()) {
+    Chirp curr = queue.front();
+    auto chirpq = reply->add_chirps();
+    chirpq[index] = curr;
+    auto replies = GetIdReply(curr.id());
+    if (replies.size() > 0) {
+      for (const auto &reply : replies) {
+        string chirpstr = GetIdChirp(reply);
+        Chirp *chirp = helper::StringToChirp(chirpstr);
```
Comment (barath) [inline service/service.cc]: Who owns this memory?

File: service/service.h [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                FollowReply *reply) override;
+  // read the whole chirp thread given chirp id through breadth first search
+  Status read(ServerContext *context, const ReadRequest *request,
+              ReadReply *reply) override;
+  // monitor the chirp sent by followed user.
+  Status monitor(ServerContext *context, const MonitorRequest *request,
+                 ServerWriter<MonitorReply> *reply) override;
+
+private:
+  IdGenerator idG;
```
Comment (barath) [inline service/service.h]: Rename to follow convention.

**Pull request merged**